Libraries/Lists/CellRenderMask.js

```diff
@@ -0,0 +1,145 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @flow strict
+ * @format
+ */
+
+import invariant from 'invariant';
+
+export type CellRegion = {
+  first: number,
+  last: number,
+  isSpacer: boolean,
+};
+
+export class CellRenderMask {
+  _numCells: number;
+  _regions: Array<CellRegion>;
+
+  constructor(numCells: number) {
+    invariant(
+      numCells >= 0,
+      'CellRenderMask must contain a non-negative number os cells',
+    );
+
+    this._numCells = numCells;
+
+    if (numCells === 0) {
+      this._regions = [];
+    } else {
+      this._regions = [
+        {
+          first: 0,
+          last: numCells - 1,
+          isSpacer: true,
+        },
+      ];
+    }
+  }
+
+  enumerateRegions(): $ReadOnlyArray<CellRegion> {
+    return this._regions;
+  }
+
+  addCells(cells: {first: number, last: number}) {
```

Comment:
The current VirtualizedList implementation I have in https://github.com/facebook/react-native/pull/31422 will compute a whole new mask if inputs change, instead of trying to reason about the existing one.

A cell can have multiple conditions that would cause it to be rendered, so we do not know at the absense of just one condition that the cell should not be rendered. So it becomes tricky to mutate the existing mask.

Reply:
Is there a use case of adding cells that are spacers?

Reply:
The render function in the linked PR above gives a good example of usage. When we hit spacers we still need to add views to the scroll view. So this structure allows us to iterate through regions and either add cells, or if a spacer, add the spacer view. If we omitted spacers we would need to derive where they were in VirtualizedList.